cpp/src/arrow/dataset/file_rados_parquet.h

```diff
@@ -0,0 +1,182 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+// This API is EXPERIMENTAL.
+#define _FILE_OFFSET_BITS 64
+
+#pragma once
+
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+#include <functional>
+#include <memory>
+#include <sstream>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "arrow/api.h"
+#include "arrow/compute/exec/expression.h"
+#include "arrow/dataset/dataset.h"
+#include "arrow/dataset/discovery.h"
+#include "arrow/dataset/file_base.h"
+#include "arrow/dataset/file_parquet.h"
+#include "arrow/dataset/rados.h"
+#include "arrow/dataset/scanner.h"
+#include "arrow/dataset/type_fwd.h"
+#include "arrow/dataset/visibility.h"
+#include "arrow/filesystem/api.h"
+#include "arrow/filesystem/path_util.h"
+#include "arrow/io/api.h"
+#include "arrow/ipc/api.h"
+#include "arrow/util/iterator.h"
+#include "arrow/util/macros.h"
+
+#include "parquet/arrow/writer.h"
+#include "parquet/exception.h"
+
+namespace arrow {
+namespace dataset {
+
+class ARROW_DS_EXPORT RadosCluster {
+ public:
+  struct RadosConnectionCtx {
+    std::string ceph_config_path;
+    std::string data_pool;
+    std::string user_name;
+    std::string cluster_name;
+    std::string cls_name;
+  };
+  explicit RadosCluster(RadosConnectionCtx& ctx)
+      : ctx(ctx), rados(new RadosWrapper()), ioCtx(new IoCtxWrapper()) {}
+
+  ~RadosCluster() { Shutdown(); }
+
+  Status Connect() {
+    if (rados->init2(ctx.user_name.c_str(), ctx.cluster_name.c_str(), 0))
+      return Status::Invalid("librados::init2 returned non-zero exit code.");
```

Comment:
It might be simpler if `RadosInterface` returned `Status` instead of `int`.  For example, the compression_xyz.cc files convert from libxyz to Status.  You could also follow the pattern used in hdfs where `hdfs_internal.h` returns int error codes but then it is hidden by `hdfs.h` which exposes everything with `Status`/`Result`.